video/out/vo_sixel.c

```diff
@@ -0,0 +1,483 @@
+#define ESC_CLEAR_COLORS "\033[0m"
+#define ESC_GOTOXY "\033[%d;%df"
+#define ESC_COLOR_BG "\033[48;2;%d;%d;%dm"
+#define ESC_COLOR_FG "\033[38;2;%d;%d;%dm"
+#define ESC_COLOR256_BG "\033[48;5;%dm"
+#define ESC_COLOR256_FG "\033[38;5;%dm"
+#define DEFAULT_WIDTH 80
+#define DEFAULT_HEIGHT 25
+
+struct vo_sixel_opts {
+    int diffuse;
+    int width;
+    int height;
+    int reqcolors;
+    int fixedpal;
+    int threshold;
+    int top;
+    int left;
+};
+
+#define OPT_BASE_STRUCT struct vo_sixel_opts
+static const struct m_sub_options vo_sixel_conf = {
+    .opts = (const m_option_t[]) {
+        {"vo-sixel-dither-algo", OPT_CHOICE(diffuse,
+            {"auto", DIFFUSE_AUTO},
+            {"none", DIFFUSE_NONE},
+            {"atkinson", DIFFUSE_ATKINSON},
+            {"fs", DIFFUSE_FS},
+            {"jajuni", DIFFUSE_JAJUNI},
+            {"stucki", DIFFUSE_STUCKI},
+            {"burkes", DIFFUSE_BURKES})},
+        {"vo-sixel-width", OPT_INT(width)},
+        {"vo-sixel-height", OPT_INT(height)},
+        {"vo-sixel-reqcolors", OPT_INT(reqcolors)},
+        {"vo-sixel-fixedpallete", OPT_INT(fixedpal)},
+        {"vo-sixel-color-threshold", OPT_INT(threshold)},
+        {"vo-sixel-offset-top", OPT_INT(top)},
+        {"vo-sixel-offset-left", OPT_INT(left)},
+        {0}
+    },
+    .defaults = &(const struct vo_sixel_opts) {
+        .diffuse = DIFFUSE_ATKINSON,
+        .width = 320,
+        .height = 240,
+        .reqcolors = 256,
+        .fixedpal = 0,
+        .threshold = 0,
+        .top = 0,
+        .left = 0,
+    },
+    .size = sizeof(struct vo_sixel_opts),
+};
+
+struct priv {
+    struct vo_sixel_opts *opts;
+
+    sixel_output_t *output;
+    sixel_dither_t *dither;
+    sixel_dither_t *testdither;
+    uint8_t        *buffer;
+
+    int image_height;
+    int image_width;
+    int image_format;
+    int screen_width;
+    int screen_height;
+
+    struct mp_image *frame;
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_sws_context *sws;
+};
+
+static const unsigned int depth = 3;
+
+static void scroll_on_demand(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    struct winsize size = {0, 0, 0, 0};
+    struct termios old_termios;
+    struct termios new_termios;
+    FILE* sixel_output_file = stdout;
+    int top = 0;
+    int left = 0;
+    int cellheight;
+    int scroll;
+    fd_set rfds;
+    struct timeval tv;
+    int ret = 0;
+
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &size);
+    if (size.ws_ypixel <= 0) {
+        fprintf(sixel_output_file, "\033[H\0337");
+        return;
+    }
+    /* set the terminal to cbreak mode */
+    tcgetattr(STDIN_FILENO, &old_termios);
+    memcpy(&new_termios, &old_termios, sizeof(old_termios));
+    new_termios.c_lflag &= ~(ECHO | ICANON);
+    new_termios.c_cc[VMIN] = 1;
+    new_termios.c_cc[VTIME] = 0;
+    tcsetattr(STDIN_FILENO, TCSAFLUSH, &new_termios);
+
+    /* request cursor position report */
+    fprintf(sixel_output_file, "\033[6n");
```

Comment:
Whenever you use an escpe sequence at the code, please define it at the top of the file, similar to `ESC_HIDE_CURSOR` (fixed string) or `ESC_GOTOXY` (parametric).

Reply:
Done